compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -631,61 +636,78 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             substs: tcx.mk_substs_trait(ty, &[]),
         };
 
-        let unpin_obligation = Obligation::new(
+        let unpin_obligation_root = Obligation::new(
             cause.clone(),
             self.param_env,
             unpin_trait_ref.without_const().to_predicate(tcx),
         );
 
-        if self.infcx.predicate_may_hold(&clone_obligation)
-            && self.infcx.predicate_may_hold(&sync_obligation)
-            && self.infcx.predicate_may_hold(&unpin_obligation)
-        {
-            return None
-        }
-
+        // Check whether catpured fields also implement the trait
         let mut temp_reasons = FxHashSet::default();
 
-        if !self.infcx.predicate_may_hold(&clone_obligation) {
-            temp_reasons.insert("Clone");
-        }
+        for capture in root_var_min_capture_list.iter() {
+            let ty = capture.place.ty();
 
-        if !self.infcx.predicate_may_hold(&sync_obligation) {
-            temp_reasons.insert("Sync");
-        }
+            // <ty as Clone>
+            let clone_trait_ref = TraitRef {
+                def_id: tcx.lang_items().clone_trait().unwrap(),
+                substs: tcx.mk_substs_trait(ty, &[]),
+            };
 
-        if !self.infcx.predicate_may_hold(&unpin_obligation) {
-            temp_reasons.insert("Unpin");
-        }
+            let clone_obligation = Obligation::new(
+                cause.clone(),
+                self.param_env,
+                clone_trait_ref.without_const().to_predicate(tcx),
+            );
 
-        let root_var_min_capture_list = if let Some(root_var_min_capture_list) =
-            min_captures.and_then(|m| m.get(&var_hir_id))
-        {
-            root_var_min_capture_list
-        } else {
-            // The upvar is mentioned within the closure but no path starting from it is
-            // used.
+            // <ty as Sync>
```

Comment:
I can't figure out how we can check for Send trait? Maybe I am missing where it is suppose to be??